_posts/2018-04-20-totw-120.md

```diff
@@ -0,0 +1,171 @@
+is why behavior is unspecified.
+
+This tip only applies when you are returning a non-reference local variable.
+Returning any other expression will not trigger this problem.
+
+## The Problem
+
+There are 2 different optimizations on return statements that can modify the
+behavior of that code snippet: NRVO (See [TotW #11](/tips/11)) and implicit
+moves.
+
+The *before* code was working because copy elision is taking place and the
+`return` statement is not actually doing any work. The variable `status` is
+already constructed in the return address and the cleanup object is seeing this
+unique instance of the `MyStatus` object _after_ the return statement.
+
+In the *after* code, copy elision is not being applied and the returned variable
+is being moved into the return value. The cleanup object is being run after the
+move operation is done and it is seeing a moved-from `MyStatus` object.
+
+Note that the *before* code was not correct either, as it relies on the copy
+elision optimization for correctness. We encourage you to rely on copy elision
+for performance (See [TotW #24](/tips/24)), but not correctness. After all,
+copy elision is an _optional_ optimization and compiler options or quality of
+implementation of the compiler can affect whether or not it happens.
+
+## Solution
+
+Do not touch the return variable after the return statement. Be careful of
+destructors of local variables doing so implicitly.
+
+The simplest solution is to separate the function in two. One that does all the
+processing, and one that calls the first one and does the post-processing (ie
+logging on error). Eg:
+
+```c++
+MyStatus DoSomething() {
+  MyStatus status;
+  status = DoA();
+  if (!status.ok()) return status;
+  status = DoB();
+  if (!status.ok()) return status;
+  status = DoC();
+  if (!status.ok()) return status;
+  return status;
+}
+MyStatus DoSomethingAndLog() {
+  MyStatus status = DoSomething();
+  if (!status.ok()) LOG(ERROR) << status;
+  return status;
+}
+```
+
+If you are only reading the value, you could also make sure to disable the
+optimizations instead. That would force a copy to be made all the time and the
+post-processing will not see a moved-from value. E.g.:
+
+```c++
+MyStatus DoSomething() {
+  MyStatus status_no_nrvo;
+  // 'status' is a reference so NRVO and all associated optimizations
+  // will be disabled.
+  // The 'return status;' statements will always copy the object and Logger
+  // will always see the correct value.
+  MyStatus& status = status_no_nrvo;
+  auto log_on_error = gtl::MakeCleanup([&status] {
+    if (!status.ok()) LOG(ERROR) << status;
+  });
+  status = DoA();
+  if (!status.ok()) return status;
+  status = DoB();
+  if (!status.ok()) return status;
+  status = DoC();
+  if (!status.ok()) return status;
+  return status;
+}
+```
+
+## Another real life example
+
+```c++
+string EncodeVarInt(int i) {
+  string out;
+  proto2::io::StringOutputStream string_output(&out);
```

Comment:
I'd just drop the namespaces, since these aren't meaningful APIs externally anyway.